### PR TITLE
ignore meson subproject directories

### PR DESCRIPTION
**Pull request by ventosus**
Meson subprojects should be handled similarly to git submodules. Meson
subprojects may actually be git submodules, but they may also just be
regular directories e.g. included via git subtree or via other means.

If the project of question uses meson as its build system, e.g. has the
following file:

    ./meson.build

Then directories in the subprojects root directory should be ignored by
default:

    ./subprojects/libfoo/foo.c
    ./subprojects/libbar/bar.c

Normal files in the subprojects root directory are normally parsed:

    ./subprojects/libfoo.wrap
    ./subprojects/libbar.wrap

This can be overriden with the command line switch:

    reuse --include-meson-subprojects lint

Relevant resources are at [1] and [2].

[1] https://mesonbuild.com/Subprojects.html
[2] https://mesonbuild.com/Wrap-dependency-system-manual.html

### CODE REVIEW

**Comment by mxmehl** [comment]
Thanks for your suggestion and the code!

Before looking at the code too closely (for which others are better suited anyway), I am not sure whether we want to have this flag enabled by default like the submodules. I assume there are other projects out there that contain a `subproject` directory without using meson at all.

I comparison, Git submodules are excluded by default by the REUSE spec, so it makes totally sense to also exclode them in the tool.

File: tests/test_main.py [hunk 1]
```diff
@@ -93,6 +93,97 @@ def test_lint_submodule_included_fail(submodule_repository, stringio):
     assert ":-(" in stringio.getvalue()
 
 
+def test_lint_meson_subprojects(fake_repository, stringio):
+    """Run a successful lint."""
+    (fake_repository / "meson.build").write_text(
+        "SPDX-FileCopyrightText: 2022 Jane Doe\n"
```
Comment (mxmehl) [inline tests/test_main.py:99]: This probably throws a number of lint errors. Please escape it in some way, e.g. by splitting the string or making it lowercase (and uppercase later). There are a number of examples in other tests.

Also applies to lines below, e.g. 108/109.

**Review by mxmehl — CHANGES_REQUESTED** [review]

File: tests/test_main.py (hunk 1 above)
Comment (ventosus) [reply tests/test_main.py:99]: Sorry, I've no idea what the problem is, could you be more specific?

    pytest -k meson

runs through without errors...

**Comment by ventosus** [comment]
> I assume there are other projects out there that contain a `subproject` directory without using meson at all.

The code of course checks whether meson is used as build system, and only then excludes directories in the *subprojects* folder.

> I comparison, Git submodules are excluded by default by the REUSE spec, so it makes totally sense to also exclode them in the tool.

Maybe I'm blind, but I cannot see any mention of git submodules in the spec to be excluded from linting Only files that are part of version control and specifically excluded from version control are skipeed while linting. git submodules *are* under version control.

File: tests/test_main.py (hunk 1 above)
Comment (mxmehl) [reply tests/test_main.py:99]: Yes, but running `reuse lint` over the project fails. This is also shown in the Actions (also `black` seems to demand changes).

You could escape it by this:
https://github.com/fsfe/reuse-tool/blob/7ac2cd03b18a45463ae353763c7ef43757f510b6/tests/test_project.py#L106-L108

**Comment by mxmehl** [comment]
> The code of course checks whether meson is used as build system, and only then excludes directories in the _subprojects_ folder.

Ah OK, I see. @carmenbianca @silverhook what do you think about such a specific ignore?

> Maybe I'm blind, but I cannot see any mention of git submodules in the spec to be excluded from linting .Only files that are part of version control or specifically excluded from version control are skipeed while linting. git submodules _are_ under version control.

You are right, this issue is still open but the path is clear: the spec will ignore Git submodules. fsfe/reuse-docs#36

**Comment by ventosus** [comment]
> You are right, this issue is still open but the path is clear: the spec will ignore Git submodules. fsfe/reuse-docs#36

I see. This solves the issue for those that include subprojects via git submodules but not for those that include subprojects via other means, e.g. git subtree or verbatim copy.

For a meson project it's pretty clear that every directory und ./subprojects *is* a subproject. If the *--recursive* flag will ever come around, it could make sense to subsum meson subprojects with git submodules.

Why have I come up with this patch in the first place?

Really a lot of my projects include subprojects via *git submodule* and  look like this, e.g. they're all esoteric in @carmenbianca's terms ;-)

*projA
 * meson.build
 * LICENSES
  * MIT 
 * src
  * foo.c 
 * subprojects
  * projA 
   * meson.build
   * LICENSES
    * MIT
    * ISC
   * .reuse
    * dep5 (Files ext / * ISC)
   * src
    * bar.c
   * ext
    * hi.c
As the subprojects are mine, too, the parent project would comply to REUSE, but in order to do so, I need to duplicate the dep5 files from all subprojects which is superfluous, as the information is already there.

Maybe there's a better solution for my use case than this patch here? I don't know.

**Comment by ventosus** [comment]
As a side note, all my subprojects are checked for and pass REUSE compatibility by their own as part of the build/test mechanism, so no information is lost by excluding them from linting of the parent project.

File: tests/test_main.py (hunk 1 above)
Comment (carmenbianca) [reply tests/test_main.py]: Fixed in [78e172d](https://github.com/fsfe/reuse-tool/pull/496/commits/78e172d02674745edeb1212b6ff904ff3a8750de)

**Comment by carmenbianca** [comment]
Pushed some changes/fixes to the branch; now LGTM.

Thanks @ventosus !

Ready for review.

**Comment by carmenbianca** [comment]
Ah, this still needs a change log entry.

**Review by mxmehl — APPROVED** [review]
LGTM, but yes, please do the changelog entry before merging.

**Comment by seabass-labrax** [comment]
Thank you! I'm looking forward to this change :)

**Pull request merged**